### PR TITLE
[5.8][stdlib] Speed up short UTF-16 distance calculations

### PR DESCRIPTION
**Pull request by lorentey**
Cherry pick of #62717 for 5.8.

Previously we insisted on using UTF-16 breadcrumbs even if we only needed to travel a very short way. This could be as much as ~ten~ forty times slower than the naive algorithm of simply visiting all the Unicode scalars in between the start and the end.

(Using breadcrumbs generally means that we need to walk to both endpoints from their nearest breadcrumb, which on average requires walking half the distance between breadcrumbs, twice — and this can mean visiting vastly more Unicode scalars than if we simply walked through the ones that are lying in between the endpoints themselves.)

To put it another way, when we want to measure how long it takes to walk between two trees within a nearby park, it probably isn't a great idea to start by separately measuring each of their distances from the nearest airport. 😛

rdar://103575481

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci test

**Review by stephentyrone — APPROVED** [review]
